sql/core/src/main/java/org/apache/spark/sql/sources/v2/Table.java

```diff
@@ -43,17 +41,8 @@
   String name();
 
   /**
-   * Returns the schema of this table.
+   * Returns the schema of this table. If the table is not readable and doesn't have a schema, an
+   * empty schema can be returned here.
    */
   StructType schema();
```

Comment:
Eh, sorry if I missed something. I guess `SupportsBatchWrite` is also going to extend `Table` .. Am I correct?

Let's suppose one datasource implements both read support and write support. How the implemented datasource knows we're in read side or write side when `schema` is called?

The source might not be able to know the schema, or fetching schema can be potentially expensive. We need a way to distinguish. Otherwise, the interface will force to read the schema everytime, which might not be able every time, or potentially expensive.

Reply:
It's also weird that write side needs a schema in a way because we need it only when Append mode (correct me if I missed some discussions or changes). Sounds like we're adding `schema` for one exception.

Reply:
I think schema is a property of table, it should be consistent between read and write.

But you do have a point that, fetching schema can be potentially expensive, and we shouldn't do it at write side if validation is not needed.

We discussed this in the data source v2 community meeting today. We will introduce a capability API, so that data source can tell Spark it doesn't want to do validation, and Spark would not call `schema()`.

Does it sound good to you?

Reply:
Ahh, I regret that I missed today's meeting. Sounds good to me and that exactly addresses my concern.